datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -25,9 +25,144 @@ use arrow::datatypes::{
+            rhs: t.clone(),
+            ret: t,
+        }
+    }
+
+    /// A signature where the inputs are coerced to the same type with a boolean output
+    fn comparison(t: DataType) -> Self {
+        Self {
+            lhs: t.clone(),
+            rhs: t,
+            ret: DataType::Boolean,
+        }
+    }
+}
+
+/// Returns a [`Signature`] for applying `op` to arguments of type `lhs` and `rhs`
+fn signature(lhs: &DataType, op: &Operator, rhs: &DataType) -> Result<Signature> {
+    match op {
+        Operator::Eq |
+        Operator::NotEq |
+        Operator::Lt |
+        Operator::LtEq |
+        Operator::Gt |
+        Operator::GtEq |
+        Operator::IsDistinctFrom |
+        Operator::IsNotDistinctFrom => {
+            comparison_coercion(lhs, rhs).map(Signature::comparison).ok_or_else(|| {
+                DataFusionError::Plan(format!(
+                    "Cannot infer common argument type for comparison operation {lhs} {op} {rhs}"
+                ))
+            })
+        }
+        Operator::And | Operator::Or => match (lhs, rhs) {
+            // logical binary boolean operators can only be evaluated in bools or nulls
+            (DataType::Boolean, DataType::Boolean)
+            | (DataType::Null, DataType::Null)
+            | (DataType::Boolean, DataType::Null)
+            | (DataType::Null, DataType::Boolean) => Ok(Signature::coerced(DataType::Boolean)),
+            _ => Err(DataFusionError::Plan(format!(
+                "Cannot infer common argument type for logical boolean operation {lhs} {op} {rhs}"
+            ))),
+        },
+        Operator::RegexMatch |
+        Operator::RegexIMatch |
+        Operator::RegexNotMatch |
+        Operator::RegexNotIMatch => {
+            regex_coercion(lhs, rhs).map(Signature::comparison).ok_or_else(|| {
+                DataFusionError::Plan(format!(
+                    "Cannot infer common argument type for regex operation {lhs} {op} {rhs}"
+                ))
+            })
+        }
+        Operator::BitwiseAnd
+        | Operator::BitwiseOr
+        | Operator::BitwiseXor
+        | Operator::BitwiseShiftRight
+        | Operator::BitwiseShiftLeft => {
+            bitwise_coercion(lhs, rhs).map(Signature::coerced).ok_or_else(|| {
+                DataFusionError::Plan(format!(
+                    "Cannot infer common type for bitwise operation {lhs} {op} {rhs}"
+                ))
+            })
+        }
+        Operator::StringConcat => {
+            string_concat_coercion(lhs, rhs).map(Signature::coerced).ok_or_else(|| {
+                DataFusionError::Plan(format!(
+                    "Cannot infer common string type for string concat operation {lhs} {op} {rhs}"
+                ))
+            })
+        }
+        Operator::Plus |
+        Operator::Minus |
+        Operator::Multiply |
+        Operator::Divide|
+        Operator::Modulo =>  {
+            // TODO: this logic would be easier to follow if the functions were inlined
```

Comment:
I plan to do this a follow up PR, but wanted to avoid conflating too many things in one PR